tuplex/core/src/ee/local/LocalBackend.cc

```diff
@@ -727,6 +742,86 @@ namespace tuplex {
         return pip_object;
     }
 
+    std::vector<std::tuple<size_t, PyObject*>> inputExceptionsToPythonObjects(std::vector<Partition *> partitions, Schema schema) {
+        using namespace tuplex;
+
+        std::vector<std::tuple<size_t, PyObject*>> pyObjects;
+        for (auto &partition : partitions) {
+            auto numRows = partition->getNumRows();
+            const uint8_t* ptr = partition->lock();
+
+            for (int i = 0; i < numRows; ++i) {
+                int64_t rowNum = *((int64_t*)ptr);
+                ptr += sizeof(int64_t);
+                int64_t ecCode = *((int64_t*)ptr);
+                ptr += 2 * sizeof(int64_t);
+                int64_t objSize = *((int64_t*)ptr);
+                ptr += sizeof(int64_t);
+
+                PyObject* pyObj = nullptr;
+                python::lockGIL();
+                if (ecCode == ecToI64(ExceptionCode::PYTHON_PARALLELIZE)) {
+                    pyObj = python::deserializePickledObject(python::getMainModule(), (char *) ptr, objSize);
+                } else {
+                    pyObj = python::rowToPython(Row::fromMemory(schema, ptr, objSize), true);
+                }
+                python::unlockGIL();
+
+                ptr += objSize;
+                pyObjects.emplace_back(rowNum, pyObj);
+            }
+
+            partition->unlock();
+            partition->invalidate();
+        }
+
+        return pyObjects;
+    }
+
+    void setExceptionInfo(const std::vector<Partition*> &normalOutput, const std::vector<Partition*> &exceptions, std::unordered_map<std::string, std::tuple<size_t, size_t, size_t>> &partitionToExceptionsMap) {
+        if (exceptions.empty()) {
+            for (const auto &p : normalOutput) {
+                partitionToExceptionsMap[uuidToString(p->uuid())] = std::make_tuple(0, 0, 0);
+            }
+            return;
+        }
+
+        auto expRowCount = 0;
+        auto expInd = 0;
+        auto expOff = 0;
+        auto expNumRows = exceptions[0]->getNumRows();
+        auto expPtr = exceptions[0]->lockWrite();
+        auto rowsProcessed = 0;
+        for (const auto &p : normalOutput) {
+            auto pNumRows = p->getNumRows();
```

Comment:
is the code within the loop exception-safe? I.e. nowhere a throw is produced? If not, this could lead to a hard to detect deadlock